cl/search/tasks.py

```diff
@@ -1770,6 +1792,55 @@ def get_embeddings_cache_key(batch_uuid: str, batch_range: str) -> str:
     return f"{embeddings_cache_key()}{batch_uuid}-{batch_range}"
 
 
+@app.task(
+    bind=True,
+    autoretry_for=(
+        NetworkError,
+        TimeoutException,
+        RemoteProtocolError,
+        HTTPStatusError,
+        ReadError,
+    ),
+    max_retries=5,
+    retry_backoff=10,
+)
+def compute_single_opinion_embeddings(self, pk: int) -> None:
+    """
+    Celery task to compute and cache text embeddings for a single opinion.
+
+    This task fetches the opinion record, extracts its best available text,
+    and sends it to the `inception-text` microservice to generate embeddings.
+    If embeddings are successfully computed, the results are cached for 30
+    minutes.
+
+    Automatically retries up to 5 times with exponential backoff (10s base)
+    in case of network or protocol-related errors.
+
+    :param self: The Celery task.
+    :param pk: The primary key of the opinion to process.
+    """
+    opinion = Opinion.objects.filter(id=pk).with_best_text().first()
+    if not opinion:
+        return None
+
+    embeddings = asyncio.run(
+        microservice(
+            service="inception-text",
+            data=opinion.clean_text,
+        )
+    )
+    # Exit early if the microservice call failed
+    if not embeddings.is_success:
```

Comment:
Sometimes when Inception fails to generate embeddings, the result is not a list but a dictionary containing the response error.

I think it’s a good idea to handle this case so we don’t try to index anything that isn’t actually embeddings, and so we can acknowledge the error for debugging.

We can do something similar to what we did in the save_embeddings task:

    if not isinstance(embeddings, list):
        if isinstance(embeddings, dict):
            logger.error(
                "Received API error response in embeddings: %s",
                json.dumps(embeddings, default=str),
            )
        else:
            logger.error(
                "Unexpected data type for embeddings: %s (%s)",
                str(embeddings)[:200],
                type(embeddings),
            )

Reply:
Thanks for the suggestion. I'll update the task

Reply:
Great! I was wondering if we can avoid this duplicated code, which is also present in save_embeddings. Maybe we could create a method to log the errors and reuse it in both places?